Signal/src/util/OWSBackupExportJob.m

```diff
@@ -211,310 +369,297 @@ - (void)configureExport:(OWSBackupJobBoolCompletion)completion
         }];
 }
 
-- (void)exportDatabase:(OWSBackupJobBoolCompletion)completion
+- (BOOL)exportDatabase
 {
-    OWSAssert(completion);
+    OWSAssert(self.backupIO);
 
     DDLogVerbose(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
 
-    if (![OWSBackupJob generateRandomDatabaseKeySpecWithKeychainKey:kOWSBackup_ExportDatabaseKeySpec]) {
-        OWSProdLogAndFail(@"%@ Could not generate database key spec for export.", self.logTag);
-        return completion(NO);
-    }
+    [self updateProgressWithDescription:NSLocalizedString(@"BACKUP_EXPORT_PHASE_DATABASE_EXPORT",
+                                            @"Indicates that the database data is being exported.")
+                               progress:nil];
 
-    // TODO: Move this work into the "export database" method.
-    NSString *jobDatabaseDirPath = [self.jobTempDirPath stringByAppendingPathComponent:@"database"];
-    if (![OWSFileSystem ensureDirectoryExists:jobDatabaseDirPath]) {
-        OWSProdLogAndFail(@"%@ Could not create jobDatabaseDirPath.", self.logTag);
-        return completion(NO);
+    YapDatabaseConnection *_Nullable dbConnection = self.primaryStorage.newDatabaseConnection;
+    if (!dbConnection) {
+        OWSProdLogAndFail(@"%@ Could not create dbConnection.", self.logTag);
+        return NO;
     }
 
-    BackupStorageKeySpecBlock keySpecBlock = ^{
-        NSData *_Nullable databaseKeySpec =
-            [OWSBackupJob loadDatabaseKeySpecWithKeychainKey:kOWSBackup_ExportDatabaseKeySpec];
-        if (!databaseKeySpec) {
-            OWSProdLogAndFail(@"%@ Could not load database keyspec for export.", self.logTag);
-        }
-        return databaseKeySpec;
+    OWSDBExportStream *exportStream = [[OWSDBExportStream alloc] initWithBackupIO:self.backupIO];
+
+    __block BOOL aborted = NO;
+    typedef BOOL (^EntityFilter)(id object);
+    typedef NSUInteger (^ExportBlock)(YapDatabaseReadTransaction *,
```

Comment:
I'm never quite sure if this "private block" pattern is more readable than a class method.